autonomi/src/client/high_level/vault/mod.rs

```diff
@@ -78,21 +144,203 @@ impl Client {
         content_type: VaultContentType,
     ) -> Result<AttoTokens, VaultError> {
-        let scratch_address = ScratchpadAddress::new(secret_key.public_key());
-        info!("Writing to vault at {scratch_address:?}");
-
-        match self
-            .scratchpad_update(secret_key, content_type, &data)
-            .await
-        {
-            Ok(()) => Ok(AttoTokens::zero()),
-            Err(ScratchpadError::CannotUpdateNewScratchpad) => {
-                let (price, _) = self
-                    .scratchpad_create(secret_key, content_type, &data, payment_option)
-                    .await?;
-                Ok(price)
+        info!("Writing {} bytes to vault ...", data.len());
+        let mut total_cost = AttoTokens::zero();
+        let main_secret_key = MainSecretKey::new(secret_key.clone());
+
+        // scratchpad_derivations ordered by the collection order
+        let (mut cur_free_graphentry_derivation, mut scratchpad_derivations) = self
+            .current_capacity(
+                &main_secret_key,
+                DerivationIndex::from_bytes(VAULT_HEAD_DERIVATION_INDEX),
+            )
+            .await?;
+
+        let content_chunks = chunk_bytes(data);
+
+        info!(
+            "Current capacity is {}, meanwhile requiring {}",
+            scratchpad_derivations.len(),
+            content_chunks.len()
+        );
+
+        // claim more capacity if short of.
+        // Note: as the Scratchpad is `created on use`, hence during the `claim stage`,
+        //       NUM_OF_SCRATCHPADS_PER_GRAPHENTRY to be claimed in one newly created GraphEntry.
+        while scratchpad_derivations.len() < content_chunks.len() {
+            let (new_free_graphentry_derivation, new_scratchpad_derivations, grap_cost) = self
+                .expand_capacity(
+                    &main_secret_key,
+                    &cur_free_graphentry_derivation,
+                    payment_option.clone(),
+                )
+                .await?;
+            cur_free_graphentry_derivation = new_free_graphentry_derivation;
+            scratchpad_derivations.extend(&new_scratchpad_derivations);
+            total_cost = AttoTokens::from_atto(total_cost.as_atto() + grap_cost.as_atto());
+        }
+
+        for (i, content) in content_chunks.into_iter().enumerate() {
+            let sp_secret_key = main_secret_key
+                .derive_key(&DerivationIndex::from_bytes(scratchpad_derivations[i].1));
+            match self
+                .scratchpad_update(&sp_secret_key.clone().into(), content_type, &content)
+                .await
+            {
+                Ok(()) => continue,
+                Err(ScratchpadError::CannotUpdateNewScratchpad) => {
+                    let (price, addr) = self
+                        .scratchpad_create(
+                            &sp_secret_key.into(),
+                            content_type,
+                            &content,
+                            payment_option.clone(),
```

Comment:
Again this would only work for wallet payment option

Reply:
same comment as at [above](https://github.com/maidsafe/autonomi/pull/2693#discussion_r1939219980)